packages/compass-editor/src/codemirror/query-history-autocompleter.ts

```diff
@@ -21,11 +21,89 @@ export type SavedQuery = {
   };
 };
 
-export const createQueryHistoryAutocompleter = (
-  savedQueries: SavedQuery[],
-  onApply: (query: SavedQuery['queryProperties']) => void,
-  theme: CodemirrorThemeType
-): CompletionSource => {
+function simplifyQueryStringForAutocomplete(queryString: string): string {
+  return queryString.replace(/[\s'"\n\t{}},]/g, '').toLowerCase();
+}
+
+/**
+ * Codemirror runs a fuzzy search on the completion item labels.
+ * Oftentimes the fuzzy search will match on too many query history items.
+ * We limit the possible results to be improve accuracy.
+ * We give suggestions of queries that either match at least one field,
+ * or that contain the prefix the user is typing.
+ */
+function getMatchingQueryHistoryItemsForInput({
+  savedQueries,
+  queryProperty,
+  input: _input,
+}: {
+  savedQueries: SavedQuery[];
+  queryProperty: string;
+  input: string;
+}) {
+  const input = simplifyQueryStringForAutocomplete(_input);
+
+  if (input.length === 0) {
+    // Everything matches when empty search.
+    return savedQueries;
+  }
+
+  return savedQueries.filter((query) => {
+    const queryValue = query.queryProperties[queryProperty];
+
+    // Only some query properties are objects. For instance limit can be an array.
+    if (typeof queryValue !== 'object') {
+      const queryValueString = toJSString(queryValue) || '';
+      if (!queryValueString) {
+        return false;
+      }
+
+      const queryValueSimplified =
+        simplifyQueryStringForAutocomplete(queryValueString);
+      return queryValueSimplified.startsWith(input);
+    }
+
+    // We subtract parts of the string until there's nothing left.
+    // We know it's a possible match if the all of the input is found in the query.
+    let inputToMatch = input;
+
+    // We check each top level field as they can be in any order.
+    for (const [key, value] of Object.entries(queryValue).slice(
+      0,
+      30 /* Some queries can have a ton of fields, we slice to avoid long loops on each character typed. */
```

Comment:
If we're introducing a performance bottleneck for rendering with this logic, we should consider making autocompleter async and moving this code out of the main thread to prevent blocking operations on main that will have a negative effect on user interactions with the editor. Let's open a ticket for that if that's the case

Reply:
Just keep in mind that async could mean that we'd have to be careful that we're not changing the results underneath the user. Like they could have started pressing enter in the meantime and then the results change again.

Reply:
@gribnoysup The occurrences of 30 fields on a query are pretty rare, I also think it's even rarer that someone would then be typing a query of that scale and then want it to auto complete. I'd like to avoid too much early optimization here, although what I'm adding here may be doing that. I'm happy to open a ticket to add async autocompletion, however I think it wouldn't be impacting many folks here. Worst case the user doesn't see a certain query autocompleted with this UI.
Do you feel async autocompletion is something we'll want to do eventually? I'm mostly wondering if we can avoid creating a ticket we might not want to ever spend the time to implement and maintain so I'd like to hear your thought.

Reply:
I'm going only based on what the comments in this code are saying: if this code is slow and because it's running every time user types any symbol in the input box, the worst case scenario is not missing items in the autocomplete, but getting a noticeable lag when typing in the input field. So if cutting the amount of fields here makes it fast enough so that it's not noticeable in terms of rendering performance, that's good enough 👍 

> Do you feel async autocompletion is something we'll want to do eventually?

I think we'll have to based on the upcoming scopes for autocompletion, but this is just me speculating at this point and should probably not affect the decision for this change 🙂

Reply:
Sounds good, I appreciate the callout there, stoked for the autocompletion project.